kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	"golang.org/x/mod/semver"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its img
+type DeploymentVersionUpdater struct {
```

Comment:
Do we still need to reconcile Deployments? 
Helm installs since Teleport 11, Teleport as a statefulset. Handling deployment -> statefulset without helm will be problematic so I guess that we should only support statefulsets by default

Reply:
That's a good question, the agent can still be deployed manually with a Deployment resource but this seems like a marginal use-case.

I don't think having a deployment controller adds much complexity as most of the logic is not deployment-specific (and StatefulSets are the painful ones to deal with, Deployments are easier). I would tend to keep the Deployment controller because of its null-ish cost but I might be underestimating the cost of maintaining it.

I suggest we keep the Deployment controller for now because it's already here and is easier to test, and decide if we should drop it once we have a working statefulset implementation and the updater is fully tested.

Reply:
Got it. 

Statefulsets are much harder to handle than Deployments.